Source/DataSources/GpxDataSource.js

```diff
@@ -0,0 +1,1029 @@
+  tmp += "background-color:" + background.toCssColorString() + ";";
+  tmp += "color:" + foreground.toCssColorString() + ";";
+  tmp += '">';
+  tmp += scratchDiv.innerHTML + "</div>";
+  scratchDiv.innerHTML = "";
+
+  //return the final HTML as the description.
+  return tmp;
+}
+
+function processWpt(dataSource, geometryNode, entityCollection, sourceUri) {
+  var position = readCoordinateFromNode(geometryNode);
+  if (!defined(position)) {
+    throw new DeveloperError("Position Coordinates are required.");
+  }
+
+  var entity = getOrCreateEntity(geometryNode, entityCollection);
+  entity.position = position;
+  // TODO different icon support
+  // var symbol = queryStringValue(geometryNode, 'sym', namespaces.gpx);
+  entity.billboard = createDefaultBillboard(dataSource._proxy, sourceUri);
+
+  var name = queryStringValue(geometryNode, "name", namespaces.gpx);
+  entity.name = name;
+  entity.label = createDefaultLabel();
+  entity.label.text = name;
+  entity.description = processDescription(geometryNode, entity);
+}
+
+//rte represents route - an ordered list of waypoints representing a series of turn points leading to a destination
+function processRte(dataSource, geometryNode, entityCollection, sourceUri) {
+  var entity = getOrCreateEntity(geometryNode, entityCollection);
+  entity.description = processDescription(geometryNode, entity);
+
+  //a list of wpt
+  var routePoints = queryNodes(geometryNode, "rtept", namespaces.gpx);
+  var coordinateTuples = new Array(routePoints.length);
+  for (var i = 0; i < routePoints.length; i++) {
+    processWpt(dataSource, routePoints[i], entityCollection, sourceUri);
+    coordinateTuples[i] = readCoordinateFromNode(routePoints[i]);
+  }
+  entity.polyline = createDefaultPolyline();
+  entity.polyline.positions = coordinateTuples;
+}
+
+//trk represents a track - an ordered list of points describing a path.
+function processTrk(dataSource, geometryNode, entityCollection, sourceUri) {
+  var entity = getOrCreateEntity(geometryNode, entityCollection);
+  entity.description = processDescription(geometryNode, entity);
+
+  var trackSegs = queryNodes(geometryNode, "trkseg", namespaces.gpx);
+  var positions = [];
+  var times = [];
+  var trackSegInfo;
+  var isTimeDynamic = true;
+  var property = new SampledPositionProperty();
+  for (var i = 0; i < trackSegs.length; i++) {
+    trackSegInfo = processTrkSeg(trackSegs[i]);
+    positions = positions.concat(trackSegInfo.positions);
+    if (trackSegInfo.times.length > 0) {
+      times = times.concat(trackSegInfo.times);
+      property.addSamples(times, positions);
+      //if one track segment is non dynamic the whole track must also be
+      isTimeDynamic = isTimeDynamic && true;
+    } else {
+      isTimeDynamic = false;
+    }
+  }
+  if (isTimeDynamic) {
+    entity.billboard = createDefaultBillboard();
+    var url = buildModuleUrl("Assets/Textures/maki/bicycle.png");
+    entity.billboard.image = resolveHref(url);
```

Comment:
I don't think `resolveHref` is actually needed, at least in how GpxDataSource is currently implemented. It is only ever called on hard-coded links to images here. If my understanding is correct, GPX doesn't actually allow for a way to link to external resources to be loaded in the application. The only links in the spec are links that can be followed (like a link to learn more about the author of the GPX file, or about this particular track etc).  

So we can simplify this code and directly use the default icons. Then the `proxy` and `sourceUri` options wouldn't be needed either. 

To replace the icons the developer can get the entities created after loading the GpxDataSource, and replace the `.image` property of the billboards, or since there's only two types of icons we can expose constructor options for them, and maybe the color of the track too.

Reply:
Removed the resolveHref proxy and sourceUri, and added options to pass wptImage (waypoint) and trkImage (track) as options, also trkColor aswell.